models/concerns/validate_data.rb

```diff
@@ -264,30 +264,76 @@ def check_trip_timewindows_consistency(relation_vehicles)
     }
   end
 
+  def check_consistent_ids_provided(relation)
+    case relation[:type]
+    when *Models::Relation::ON_VEHICLES_TYPES
+      if relation[:linked_ids].to_a.any? || relation[:linked_vehicle_ids].to_a.empty?
+        raise OptimizerWrapper::DiscordantProblemError.new(
+          "#{relation[:type]} relations do not support linked_ids and expect linked_vehicle_ids"
+        )
+      end
+    when *Models::Relation::ON_SERVICES_TYPES
+      if relation[:linked_vehicle_ids].to_a.any? || relation[:linked_ids].to_a.empty?
+        raise OptimizerWrapper::DiscordantProblemError.new(
+          "#{relation[:type]} relations do not support linked_vehicle_ids and expect linked_ids"
+        )
+      end
+    else
+      raise 'Unknown relation type' # in case we ever forget to handle a new relation type
+    end
+  end
+
+  def check_consistent_lapses_provided(relation)
+    case relation[:type]
+    when *Models::Relation::NO_LAPSE_TYPES
+      if relation[:lapse] || relation[:lapses].to_a.any?
+        raise OptimizerWrapper::DiscordantProblemError.new(
+          "#{relation[:type]} relations do not expect any lapse"
+        )
+      end
+    when *Models::Relation::ONE_LAPSE_TYPES
+      if relation[:lapse].nil? && relation[:lapses].to_a.size != 1
+        raise OptimizerWrapper::DiscordantProblemError.new(
+          "#{relation[:type]} relations expect exactly one lapse"
+        )
+      end
+    when *Models::Relation::SEVERAL_LAPSE_TYPES
+      if relation[:lapse].nil? && relation[:lapses].to_a.size != 1
+        exp_lapse_count = (relation[:linked_ids]&.size || relation[:linked_vehicle_ids]&.size) - 1
+        if relation[:lapses].to_a.size != exp_lapse_count
+          raise OptimizerWrapper::DiscordantProblemError.new(
+            "#{relation[:type]} relations expect at least one lapse"
+          ) unless relation[:type] == :vehicle_trips && relation[:lapses].to_a.empty?
```

Comment:
If I am not mistaken we already had such a logic `relation.lapse ||= 0` for other relations. I don't see why it should be a problem for `:vehicle_trips`..

This can be easily handled in the code I suggested [here](https://github.com/Mapotempo/optimizer-api/pull/265#discussion_r686970499) by adding a single line to the last `when` case

```ruby
relation[:lapses] = [0] if relation[:type] == :vehicle_trips && relation[:lapse].nil? && relation[:lapses].nil?
```
that way we don't need to care about this case rest of the code.